Casting to strptime

toanngosy opened the pull request:
Malformed timestamp failed to catch at this [line](https://github.com/AMF-FLX/AMF-BASE-QAQC/blob/develop/processing/timestamp_checks.py#L195). We don't catch the error when we can't cast the last element of TIMESTAMP_START.
The solution uses the function `cast_as_datetime()` from the `ts_util` to cast the timestamp string.
Fixes Issue #23

## PR Self Evaluation

- [x] I have reviewed my code to check that it contains no sensitive information.
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests or modified existing tests that prove my fix is effective or that my feature works
- [x] Existing unit tests pass locally with my changes

dschristianson commented:
Thanks for working on this @toanngosy.

We need an approach that catches the malformed timestamp in the `check_timestamp_format` method and thus prevents any further timestamp checks from happening.

I had a look at the code. `ts_utils.cast_as_datetime` is being used ultimately in the `check_timestamp_format` method. But the error is being logged in the wrong logger to get caught properly. Please work out an approach roughly as follows:
1) modify ts_utils.cast_as_datetime and any other methods necessary to allow the logger to be passed in. We want the fatal log message [here](https://github.com/AMF-FLX/AMF-BASE-QAQC/blob/92f0f35422b860ad11aeec1d08b3caf0692eb7d8/processing/utils.py#L109-L111) to log to the logger set up [here](https://github.com/AMF-FLX/AMF-BASE-QAQC/blob/92f0f35422b860ad11aeec1d08b3caf0692eb7d8/processing/timestamp_checks.py#L254-L257). Then, the fatal log count will set the variable ts_valid to False and the future fill check won't run.

Also, please add a test file in the ~/processing/test/testdata/format_qaqc directory that contain forward timestamps to match the use case. You'll also need to add a line to ~/processing/test/testdata/format_qaqc/file_check_expected_issues_ts3.csv for that new test file. Verify that your fix handles this test file properly -- you should get a timestamp format error and a message that timestamp checks were incomplete.

toanngosy commented:
> Thanks for working on this @toanngosy.
> 
> We need an approach that catches the malformed timestamp in the `check_timestamp_format` method and thus prevents any further timestamp checks from happening.
> 
> I had a look at the code. `ts_utils.cast_as_datetime` is being used ultimately in the `check_timestamp_format` method. But the error is being logged in the wrong logger to get caught properly. Please work out an approach roughly as follows:
> 
> 1. modify ts_utils.cast_as_datetime and any other methods necessary to allow the logger to be passed in. We want the fatal log message [here](https://github.com/AMF-FLX/AMF-BASE-QAQC/blob/92f0f35422b860ad11aeec1d08b3caf0692eb7d8/processing/utils.py#L109-L111) to log to the logger set up [here](https://github.com/AMF-FLX/AMF-BASE-QAQC/blob/92f0f35422b860ad11aeec1d08b3caf0692eb7d8/processing/timestamp_checks.py#L254-L257). Then, the fatal log count will set the variable ts_valid to False and the future fill check won't run.
> 
> Also, please add a test file in the ~/processing/test/testdata/format_qaqc directory that contain forward timestamps to match the use case. You'll also need to add a line to ~/processing/test/testdata/format_qaqc/file_check_expected_issues_ts3.csv for that new test file. Verify that your fix handles this test file properly -- you should get a timestamp format error and a message that timestamp checks were incomplete.

Thank you so much for the detailed instructions @dschristianson !

I added the test for `check_timestamp_format` where we have a bad timestamp at various positions in `TIMESTAMP_START` and `TIMESTAMP_END`.
I added some bad files related to timestamp issues (index 75 to the end) into `file_check_expected_issues_ts3.csv` and changed `run_upload_check_manual_tests.py` so it works with the current test folder structure. 
All checks from `run_upload_check_manual_tests.py` are broken because in `DataReader.read_single_file()` we no longer have `fname= and checklog=`, so I make the change here to make it works as well.

toanngosy commented:
> Thanks for working on this @toanngosy.
> 
> We need an approach that catches the malformed timestamp in the `check_timestamp_format` method and thus prevents any further timestamp checks from happening.
> 
> I had a look at the code. `ts_utils.cast_as_datetime` is being used ultimately in the `check_timestamp_format` method. But the error is being logged in the wrong logger to get caught properly. Please work out an approach roughly as follows:
> 
> 1. modify ts_utils.cast_as_datetime and any other methods necessary to allow the logger to be passed in. We want the fatal log message [here](https://github.com/AMF-FLX/AMF-BASE-QAQC/blob/92f0f35422b860ad11aeec1d08b3caf0692eb7d8/processing/utils.py#L109-L111) to log to the logger set up [here](https://github.com/AMF-FLX/AMF-BASE-QAQC/blob/92f0f35422b860ad11aeec1d08b3caf0692eb7d8/processing/timestamp_checks.py#L254-L257). Then, the fatal log count will set the variable ts_valid to False and the future fill check won't run.
> 
> Also, please add a test file in the ~/processing/test/testdata/format_qaqc directory that contain forward timestamps to match the use case. You'll also need to add a line to ~/processing/test/testdata/format_qaqc/file_check_expected_issues_ts3.csv for that new test file. Verify that your fix handles this test file properly -- you should get a timestamp format error and a message that timestamp checks were incomplete.

Thank you so much for the detailed instructions @dschristianson !

I added the test for `check_timestamp_format` where we have a bad timestamp at various positions in `TIMESTAMP_START` and `TIMESTAMP_END`.
I added some bad files related to timestamp issues (index 75 to the end) into `file_check_expected_issues_ts3.csv` and changed `run_upload_check_manual_tests.py` so it works with the current test folder structure. 
All checks from `run_upload_check_manual_tests.py` are broken because in `DataReader.read_single_file()` we no longer have `fname= and checklog=`, so I make the change here to make it works as well.
